File: a800mon/shortcutbar.py
from .app import VisualComponent
from .appstate import shortcuts, state
from .shortcuts import Shortcut
from .ui import Color


class ShortcutBar(VisualComponent):
    def __init__(self, window, shortcut_width=16, mode_width=16):
        super().__init__(window)
        self._last_mode = None
        self._shortcut_width = shortcut_width
        self._mode_width = mode_width

    def _print_slot(self, shortcut: Shortcut):
        key_text = f" {shortcut.key_as_text()} "
        label_text = " " + shortcut.label.ljust(self._shortcut_width)
        self.window.print(key_text, attr=Color.SHORTCUT.attr())
        self.window.print(label_text, attr=Color.TEXT.attr())

    def render(self, force_redraw=False):
        if force_redraw or not self._last_mode == state.active_mode:
            self.window.cursor = 0, 0

            layer = shortcuts.get(state.active_mode)
            if layer:
                self._last_mode = state.active_mode
                layer_text = layer.name[: self._mode_width].ljust(
                    self._mode_width)
                self.window.print(layer_text, layer.color.attr())

                for shortcut in layer.get_shortcuts():
                    self._print_slot(shortcut)
                self.window.fill_to_eol(attr=Color.TEXT.attr())

                globals_len = 0
                globals_list = shortcuts.global_shortcuts()
                for shortcut in globals_list:
                    globals_len += (
                        len(shortcut.key_as_text()) + 3 + self._shortcut_width
                    )
                right_start = self.window._iw - globals_len
                if right_start > 0:
                    self.window.cursor = right_start, 0
                    for shortcut in globals_list:
                        self._print_slot(shortcut)
            else:
                self._last_mode = None

File: go800mon/a800mon/monitor/shortcutbar.go
package monitor

import (
	"context"

	. "go800mon/a800mon"
)

type ShortcutBar struct {
	BaseWindowComponent
	shortcuts *ShortcutManager
	lastMode  AppMode
}

func NewShortcutBar(window *Window, shortcuts *ShortcutManager) *ShortcutBar {
	return &ShortcutBar{
		BaseWindowComponent: NewBaseWindowComponent(window),
		shortcuts:           shortcuts,
	}
}

func (s *ShortcutBar) Update(_ctx context.Context) (bool, error) {
	mode := State().ActiveMode
	if s.lastMode == mode {
		return false, nil
	}
	s.lastMode = mode
	return true, nil
}

func (s *ShortcutBar) Render(_force bool) {
	st := State()
	w := s.Window()
	w.Cursor(0, 0)
	layer := s.shortcuts.Get(int(st.ActiveMode))
	if layer != nil {
		layerText := padRight(layer.Name, 16)
		w.Print(layerText, layer.Color.Attr(), false)
		for _, sh := range layer.List() {
			s.printSlot(sh)
		}
		w.FillToEOL(' ', ColorText.Attr())
	}
	globals := s.shortcuts.Globals()
	globalW := 0
	for _, sh := range globals {
		globalW += len([]rune(sh.KeyAsText())) + 3 + 16
	}
	rightStart := w.Width() - globalW
	if rightStart > 0 {
		w.Cursor(rightStart, 0)
		for _, sh := range globals {
			s.printSlot(sh)
		}
	}
}

func (s *ShortcutBar) printSlot(shortcut Shortcut) {
	w := s.Window()
	keyText := " " + shortcut.KeyAsText() + " "
	label := " " + padRight(shortcut.Label, 16)
	w.Print(keyText, ColorShortcut.Attr(), false)
	w.Print(label, ColorText.Attr(), false)
}

func padRight(s string, n int) string {
	r := []rune(s)
	if len(r) >= n {
		return string(r[:n])
	}
	for len(r) < n {
		r = append(r, ' ')
	}
	return string(r)
}
